clang/lib/Sema/SemaConcept.cpp

```diff
@@ -1122,6 +1122,154 @@ bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+
+  bool instantiateParameterToScope(ParmVarDecl *OldParm,
+                                   LocalInstantiationScope &Scope) {
+    // The current context might have been changed by lambda expressions. So
+    // resume it before we substitute into parameters.
+    Sema::ContextRAII Context(SemaRef, PrimaryTemplatedFunction);
+    std::optional<unsigned> NumExpansions;
+    ParmVarDecl *NewParm = nullptr;
+    unsigned IndexAdjustment = 0;
+    if (OldParm->isParameterPack()) {
+      SmallVector<UnexpandedParameterPack, 2> Unexpanded;
+      TypeLoc TL = OldParm->getTypeSourceInfo()->getTypeLoc();
+      PackExpansionTypeLoc ExpansionTL = TL.castAs<PackExpansionTypeLoc>();
+      TypeLoc Pattern = ExpansionTL.getPatternLoc();
+      SemaRef.collectUnexpandedParameterPacks(Pattern, Unexpanded);
+
+      assert(!Unexpanded.empty() &&
+             "A pack Decl doesn't contain anything unexpanded?");
+
+      bool ShouldExpand = false;
+      bool RetainExpansion = false;
+      std::optional<unsigned> OrigNumExpansions =
+          ExpansionTL.getTypePtr()->getNumExpansions();
+      NumExpansions = OrigNumExpansions;
+      if (SemaRef.CheckParameterPacksForExpansion(
+              ExpansionTL.getEllipsisLoc(), Pattern.getSourceRange(),
+              Unexpanded, TemplateArgs, ShouldExpand, RetainExpansion,
+              NumExpansions))
+        return true;
+
+      assert(ShouldExpand && !RetainExpansion &&
+             "Shouldn't retain an expansion here!");
+      Scope.MakeInstantiatedLocalArgPack(OldParm);
+
+      for (unsigned I = 0; I != *NumExpansions; ++I) {
+        Sema::ArgumentPackSubstitutionIndexRAII SubstIndex(SemaRef, I);
+        ParmVarDecl *NewParm = SemaRef.SubstParmVarDecl(
+            OldParm, TemplateArgs, /*indexAdjustment=*/IndexAdjustment++,
+            NumExpansions, /*ExpectParameterPack=*/false,
+            /*EvaluateConstraints=*/false);
+        if (!NewParm)
+          return true;
+      }
+
+      return false;
+    }
+    NewParm = SemaRef.SubstParmVarDecl(OldParm, TemplateArgs,
+                                       /*indexAdjustment=*/IndexAdjustment,
+                                       std::nullopt,
+                                       /*ExpectParameterPack=*/false);
+    if (!NewParm)
+      return true;
+    Scope.InstantiatedLocal(OldParm, NewParm);
+    return false;
+  }
+
+public:
+  InstantiateReferencedParameter(
+      Sema &SemaRef, const MultiLevelTemplateArgumentList &TemplateArgs,
+      FunctionDecl *PrimaryTemplatedFunction)
+      : inherited(SemaRef), TemplateArgs(TemplateArgs),
+        PrimaryTemplatedFunction(PrimaryTemplatedFunction) {}
+
+  Decl *TransformDecl(SourceLocation Loc, Decl *D) {
+    if (auto *PVD = dyn_cast_if_present<ParmVarDecl>(D);
+        PVD && PVD->getDeclContext() == PrimaryTemplatedFunction &&
+        !InstantiatedDecls.contains(PVD)) {
+      instantiateParameterToScope(PVD, *SemaRef.CurrentInstantiationScope);
```

Comment:
We probably need to handle errors here.
But I think we should instantiate _ALL_ the parameters, not just the one referenced in the constraint

Reply:
That would cause an implementation divergence with GCC: https://gcc.godbolt.org/z/7MobE9b61
If we instantiate all parameters, we wouldn't accept such a case that GCC currently accepts.

The standard seems *vague* ([It](https://cplusplus.github.io/CWG/issues/2369.html) just said `function type` shouldn't be replaced before the constraint evaluation) about this, but I presume the intent was to instantiate parameters on demand.

Reply:
I did some research, this is https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_active.html#2770

Reply:
See also https://github.com/cplusplus/CWG/issues/364 
https://cplusplus.github.io/CWG/issues/2769.html

https://gcc.godbolt.org/z/WrfG5Gxaj

I think instead of doing the substitution there, we want to do it where we check atomic constraints

Reply:
@zygoloid

Reply:
> I think instead of doing the substitution there, we want to do it where we check atomic constraints

Err.. I just tried and I realized there will be much more complexity then:
IIUC we need to move the substitution to `TemplateInstantiator` which just delegates `TryExpandParameterPacks` to `Sema.CheckParameterPacksForExpansion`, so for a pack expansion, e.g.

```cpp
void f18(auto... x)
  requires(sizeof...(x) == 2);

void foo() {
  f18('c');
}
```

We would end up calling `CurrentInstantiationScope->findInstantiationOf` inside `CheckParameterPacksForExpansion` when trying to transform the `sizeof...` expression. Unfortunately we don't have a chance to instantiate `x` between the transform of the expression starts and the `findInstantiation` call, therefore we would crash.

One alternative is to duplicate much of the `CheckParameterPacksForExpansion` to `TryExpandParameterPacks`, and we instantiate the Decls there. But I feel that would clutter up everything and we probably couldn't accept the maintenance cost.

Reply:
Hmmm, never mind. Probably things are not so dreadful than I thought.